address_restricted_ed25519.go

```diff
@@ -0,0 +1,157 @@
+	}
+
+	ed25519Addr := Ed25519Address(addr.PubKeyHash)
+	return edSig.Valid(msg, &ed25519Addr)
+}
+
+func (addr *RestrictedEd25519Address) Equal(other Address) bool {
+	otherAddr, is := other.(*RestrictedEd25519Address)
+	if !is {
+		return false
+	}
+
+	return addr.PubKeyHash == otherAddr.PubKeyHash &&
+		bytes.Equal(addr.Capabilities, otherAddr.Capabilities)
+}
+
+func (addr *RestrictedEd25519Address) Type() AddressType {
+	return AddressRestrictedEd25519
+}
+
+func (addr *RestrictedEd25519Address) Bech32(hrp NetworkPrefix) string {
+	return bech32String(hrp, addr)
+}
+
+func (addr *RestrictedEd25519Address) String() string {
+	return hexutil.EncodeHex(lo.PanicOnErr(CommonSerixAPI().Encode(context.TODO(), addr)))
+}
+
+func (addr *RestrictedEd25519Address) Size() int {
+	return Ed25519AddressSerializedBytesSize +
+		addr.Capabilities.Size()
+}
+
+func (addr *RestrictedEd25519Address) CannotReceiveNativeTokens() bool {
+	return addr.Capabilities.CannotReceiveNativeTokens()
+}
+
+func (addr *RestrictedEd25519Address) CannotReceiveMana() bool {
+	return addr.Capabilities.CannotReceiveMana()
+}
+
+func (addr *RestrictedEd25519Address) CannotReceiveOutputsWithTimelockUnlockCondition() bool {
+	return addr.Capabilities.CannotReceiveOutputsWithTimelockUnlockCondition()
+}
+
+func (addr *RestrictedEd25519Address) CannotReceiveOutputsWithExpirationUnlockCondition() bool {
+	return addr.Capabilities.CannotReceiveOutputsWithExpirationUnlockCondition()
+}
+
+func (addr *RestrictedEd25519Address) CannotReceiveOutputsWithStorageDepositReturnUnlockCondition() bool {
+	return addr.Capabilities.CannotReceiveOutputsWithStorageDepositReturnUnlockCondition()
+}
+
+func (addr *RestrictedEd25519Address) CannotReceiveAccountOutputs() bool {
+	return addr.Capabilities.CannotReceiveAccountOutputs()
+}
+
+func (addr *RestrictedEd25519Address) CannotReceiveNFTOutputs() bool {
+	return addr.Capabilities.CannotReceiveNFTOutputs()
+}
+
+func (addr *RestrictedEd25519Address) CannotReceiveDelegationOutputs() bool {
+	return addr.Capabilities.CannotReceiveDelegationOutputs()
+}
+
+func (addr *RestrictedEd25519Address) CapabilitiesBitMask() AddressCapabilitiesBitMask {
+	return addr.Capabilities
+}
+
+// RestrictedEd25519AddressFromPubKey returns the address belonging to the given Ed25519 public key.
+func RestrictedEd25519AddressFromPubKey(pubKey ed25519.PublicKey,
+	canReceiveNativeTokens bool,
+	canReceiveMana bool,
+	canReceiveOutputsWithTimelockUnlockCondition bool,
+	canReceiveOutputsWithExpirationUnlockCondition bool,
+	canReceiveOutputsWithStorageDepositReturnUnlockCondition bool,
+	canReceiveAccountOutputs bool,
+	canReceiveNFTOutputs bool,
+	canReceiveDelegationOutputs bool) *RestrictedEd25519Address {
+
+	address := blake2b.Sum256(pubKey[:])
+	addr := &RestrictedEd25519Address{}
+	copy(addr.PubKeyHash[:], address[:])
+
+	if canReceiveNativeTokens {
```

Comment:
Should me make an utility function for the following part?
Pass the bools and return the capabilities slice?